misc/python/materialize/cli/crate_diagram.py

```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python3
+
+# Copyright Materialize, Inc. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+"""
+Generate a dependency graph of our local crates
+
+We have hundreds of crates in our dependency tree, so visualizing the full set of dependencies is
+fairly useless (but can be done with crates like cargo-graph).
+
+This script just prints the relationship of our internal crates to help see how things fit into the
+materialize ecosystem.
+"""
+
+import subprocess
+import sys
+import os
+import difflib
+from collections import defaultdict
+from pathlib import Path
+from tempfile import NamedTemporaryFile
+from typing import cast, Any, Dict, IO, List, Optional
+
+import click
+import toml
+
+from materialize.spawn import runv
+
+
+@click.command()
+@click.option("--include-timely/--no-include-timely")
+@click.option(
+    "--create-diagram",
+    default=None,
+    help="The svg diagram to create, otherwise just print dot to stdout",
+)
+@click.option("--check", default=None)
+def main(
+    include_timely: bool, create_diagram: Optional[str], check: Optional[str]
+) -> None:
+    root = Path(os.environ["MZ_ROOT"])
+    root_cargo = root / "Cargo.toml"
+    with root_cargo.open() as fh:
+        data = toml.load(fh)
+
+    members = set()
+    areas = defaultdict(list)
+    member_meta = {}
+    all_deps = {}
+    for member_path in data["workspace"]["members"]:
+        path = root / member_path / "Cargo.toml"
+        with path.open() as fh:
+            member = toml.load(fh)
+        has_bin = any(root.joinpath(member_path).glob("src/**/main.rs"))
+        name = member["package"]["name"]
+
+        member_meta[name] = {
+            "has_bin": has_bin,
+            "description": member["package"].get("description", name),
+        }
+        area = member_path.split("/")[0]
+        areas[area].append(name)
+        members.add(name)
+        all_deps[name] = [dep for dep in member.get("dependencies", [])]
+
+    if include_timely:
+        members.add("timely")
+        members.add("differential-dataflow")
+    local_deps = {
+        dep_name: [dep for dep in dep_deps if dep in members]
+        for dep_name, dep_deps in all_deps.items()
+    }
```

Comment:
A lot of this Cargo.toml parsing logic already exists in https://mtrlz.dev/api/python/materialize/cargo.html. Would be great to reuse that.

Reply:
I can ensure that this logic lives in there if it can't be reused.